[GSoC2024] add return types to fix typescript warnings

tahamukhtar20 opened the pull request:
<!-- Raise an issue to propose your change (https://github.com/opencv/cvat/issues).
It helps to avoid duplication of efforts from multiple independent contributors.
Discuss your ideas with maintainers to be sure that changes will be approved and merged.
Read the [Contribution guide](https://opencv.github.io/cvat/docs/contributing/). -->

<!-- Provide a general summary of your changes in the Title above -->
Fixes: #7598
### Motivation and context
There are some minor typescript errors in the files, this pull request fixes those.

<!-- Why is this change required? What problem does it solve? If it fixes an open
issue, please link to the issue here. Describe your changes in detail, add
screenshots. -->

### How has this been tested?
Ran the code after changing, and it was functional like before, this didn't need any additional test.
<!-- Please describe in detail how you tested your changes.
Include details of your testing environment, and the tests you ran to
see how your change affects other areas of the code, etc. -->

### Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply.
If an item isn't applicable for some reason, then ~~explicitly strikethrough~~ the whole
line. If you don't do that, GitHub will show incorrect progress for the pull request.
If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I submit my changes into the `develop` branch
- [x] I have created a changelog fragment <!-- see top comment in CHANGELOG.md -->
- [x] I have linked related issues (see [GitHub docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
  ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
  [cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning),
  [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and
  [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.

tahamukhtar20 commented:
@bsekachev I've made the necessary changes, please change the last file Lambda Manager as I have added a few Promise<any> there because it is a very reused file and to me, Promise<any> makes the most sense there. Please check and let me know if there is any issue and I'll fix it ASAP. Thanks.

tahamukhtar20 commented:
I think this PR is not ready yet, there are a few changes which I just realized need to be added, I'll add those and ping you accordingly, thanks.

bsekachev commented on cvat-core/src/lambda-manager.ts:
```diff
@@ -43,7 +43,7 @@ class LambdaManager {
         return { models, count: lambdaFunctions.length };
     }
 
-    async run(taskID: number, model: MLModel, args: any) {
```
Promise<any> should be specified more concrete in this file

bsekachev replied:
@tahamukhtar20

tahamukhtar20 replied:
I'll get to these tomorrow after submitting my proposal. Thanks for the reminder.

SpecLad commented on changelog.d/20240327_224004_tahamukhtar20.md:
```diff
@@ -0,0 +1,4 @@
+### Fixed
+
+- Fixed typescript linting errors in ts files
```
This fix has no impact on the user, so I don't think it needs a changelog entry.

bsekachev commented:
Hello, do you have any plans regarding the pull request?

tahamukhtar20 commented:
Yes @bsekachev , I left a comment above under your review, please let me know. Thanks.

bsekachev commented:
Yes, one week ago you mentioned "tomorrow". That is why I am asking :)

tahamukhtar20 commented:
> Yes, one week ago you mentioned "tomorrow". That is why I am asking :)

Sorry about that, @bsekachev  I wanted to ask something about your comment under the Promise<any> review, you said we should use something more concrete here instead of any, but that is Lambda manager, and we are very uncertain what will be returned here, hence doesn't `any` make sense here?

bsekachev commented:
I believe we are certain enough. Otherwise how would it be possible to work with this API? 
As far as I remember `run()` returns a string, representing rq identificator to fetch statuses later. 
`call()` is more complex, but, more or less, you may find it out from this file `cvat-ui/src/components/annotation-page/standard-workspace/controls-side-bar/tools-control.tsx`. 

Perhaps the output is different for trackers, detectors and interactors. 
So, it should be like `Promise<TrackerResults | DetectorResults | InteractorResults>`, and define these three types additionally.

tahamukhtar20 commented:
Hi @bsekachev, I've added the required types and currently added the interfaces on top of the Lambda Manager file, please let me know if that is fine or if there is some other place more appropriate for the interfaces.
Additionally, I have updated the concerned code related to the lambda.call function, but I had to make some type assertions(e.g. number[] and number[][] in the tools-control.tsx file) to make it work with the existing code without making too many changes. Please let me know if this is the correct implementation for this.
Thank you.

bsekachev commented on cvat-core/src/lambda-manager.ts:
```diff
@@ -6,14 +6,36 @@
 
+export interface InteractorResults {
+    mask: number[][];
+    points: number[][] | number[];
+    bounds: [number, number, number, number]
```
```suggestion
    points?: [number, number][];
    bounds?: [number, number, number, number]
```

bsekachev commented on cvat-core/src/lambda-manager.ts:
```diff
@@ -6,14 +6,36 @@
+}
+
+export interface TrackerResults {
+    states: any[]
+    shapes: number[][]
```
```suggestion
    states: any[];
    shapes: number[][];
```

bsekachev commented on cvat-core/src/lambda-manager.ts:
```diff
@@ -6,14 +6,36 @@
+    bounds: [number, number, number, number]
+}
+
+export interface SerializedShape {
+    type: ShapeType;
+    rotation?: number;
+    attributes: { name: string; value: string }[];
+    label: string;
+    outside?: boolean;
+    points?: number[];
+    mask?: number[];
+    elements: SerializedShape[];
+}
```
```suggestion
export interface DetectedShape {
    type: ShapeType;
    rotation?: number;
    attributes: { name: string; value: string }[];
    label: string;
    outside?: boolean;
    points?: number[];
    mask?: number[];
    elements: DetectedShape[];
}
```

bsekachev commented on cvat-core/src/lambda-manager.ts:
```diff
@@ -68,7 +90,7 @@ class LambdaManager {
     }
 
-    async call(taskID, model, args) {
+    async call(taskID, model, args): Promise<TrackerResults | InteractorResults | SerializedShape[]> {
```
```suggestion
    async call(taskID, model, args): Promise<TrackerResults | InteractorResults | DetectedShape[]> {
```

bsekachev commented:
Please, avoid changes in `tools-control.tsx`, I will work with types inconsistency on my own, I see there are some issues.

tahamukhtar20 commented:
@bsekachev Updated

bsekachev commented:
Just FYI: https://github.com/cvat-ai/cvat/pull/7785

tahamukhtar20 commented:
I found some more typescript warnings because of missing return types, maybe I can create a follow-up pull request for this if you allow it.
